RFC 0715: CDK Construct Upgrade Guide (#724)
aws/aws-cdk-rfcs: text/0715-construct-upgrade-support.md

- **Original Author(s):**: @GavinZZ
- **Tracking Issue**: #715
- **API Bar Raiser**: @iliapolo

A construct upgrade framework in the CDK CLI that helps developers upgrade from deployed and
older construct libraries to modern, fully supported alternatives that incorporate the latest
features and best practices.

## Motivation

As the AWS Cloud Development Kit (CDK) evolves, we continuously deliver new features,
performance optimizations, and reliability improvements. To maintain backward compatibility
while enabling forward progress, the CDK team uses structured versioning strategies for critical updates.

A common pattern involves releasing versioned constructs or modules (e.g., V2 variants)
that coexist with their predecessors. For example:

* The `@aws-cdk/aws-ec2-alpha` alpha module introduces the modernized `VpcV2`, `SubnetV2`, and other
new construct, designed to succeed the original Vpc related constructs in `aws-cdk-lib/aws-ec2`.
* The `TableV2` construct in `aws-cdk-lib/aws-dynamodb` replaces legacy Table implementations
while retaining API compatibility.

These updates unlock enhanced capabilities. Construct upgrade between variants, however, often requires

> code changes that could trigger CloudFormation resource replacements (e.g., NAT gateways being

Reviewer: Internally with our own wrapper CDK we are discussing the idea of a semantic versioning schema of four entries X.X.X.X where you have major infrastructure deployment breaking change, major code breaking change, minor, patch. This allows us to decouple refactors from live service breaking change. Especially in a world with code assistants making breaking library code changes far less disruptive if non-breaking to the live resource. Just wanted to throw this out there as an idea to discuss for the aws-cdk team

blockers, such as CloudFormation drift. Additionally, it analyzes the CloudFormation change set and validates
any library-specific configurations. See [Construct Upgrade Validations](#construct-upgrade-validations) for a detailed breakdown of the validations.

In the initial release of the CDK construct upgrade tool, VPC-related constructs upgrade from `aws-cdk-lib/aws-ec2`
to `@aws-cdk/aws-ec2-alpha` and `Table` to `TableV2` in `aws-cdk-lib/aws-dynamodb` come with enhanced
construct-specific validations and comprehensive migration guides provided by the CDK team to ensure
safe deployment. These two construct upgrades are so-called `fully supported` construct upgrade.

We aim to gradually increase the number of `fully supported` constructs. The tool is also designed to be open
and pluggable, allowing users to support other construct migrations by incorporating their own custom
validations. This flexibility enables users to extend the tool to cover additional constructs that may not
yet be officially supported with custom validations, see [Custom Validations](#custom-validations).

> ## Walkthrough of Construct Upgrade Example

Reviewer 2: The example is great. I think it would also be very helpful to add the VPC example as well, since it follows the other flow (refactor).

safely migrate their infrastructure. Consider a DynamoDB table deployed to CloudFormation using the legacy
`Table` construct:

```ts
import * as dynamodb from 'aws-cdk-lib/aws-dynamodb';

export class DemoStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const table = new dynamodb.Table(this, 'MyTable', {
      partitionKey: {
        name: 'PK',
        type: dynamodb.AttributeType.STRING,
      },
      replicationRegions: ['us-west-2'],
    });
  }
}
```

CDK team provides a comprehensive migration guide that includes construct property mappings
(e.g., replicationRegions → replicas), code snippets, common migration scenarios, and etc.
To upgrade to TableV2 construct, first update your CDK code following the migration guide:

```ts
import * as dynamodb from 'aws-cdk-lib/aws-dynamodb';

export class DemoStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const table = new dynamodb.TableV2(this, 'MyTable', {
      partitionKey: {
        name: 'PK',
        type: dynamodb.AttributeType.STRING,
      },

>       tableName: 'DemoStack-MyTable794EDED1-11W4MR8VZ0UPE',

Author: This is actually not possible. The `resourcesToImport` features does not work unfortunately in this case because what it does requires the stack to be identical except the newly created resource that will be imported. What this means is that the user workflow will be torn into multiple step:
1. Users add a new TableV2 construct
2. Users run cdk import/cdk deploy (with migrate.json to use resourcesToImport feature).
3. Once import is successful, users remove the TableV1 construct
4. Deploy the change

This now becomes a multi-step and non-automated step with user intevention in between. I do not like this approach and I don't think it's worth the pain. I outlined a few possible workarounds instead to not hardcode the value in the stack.

>       tableName: 'DemoStack-MyTable794EDED1-11W4MR8VZ0UPE',

Reviewer 3: This really is a non-starter.

The reason is that this construct or stack immediately becomes bound to a single (account, region) combination, and it becomes undeployable anywhere else.

To do it "properly", this should have a lookup map:

```ts
const tableNames = {
  'us-east-1/11111111': 'DemoStack-MyTable794EDED1-11W4MR8VZ0UPE',
  'us-east-2/222222222': 'DemoStack-MyTable794EDED1-83483583'
  // ...
 };

 const table = new dynamodb.TableV2(this, 'MyTable', {
      partitionKey: {
        name: 'PK',
        type: dynamodb.AttributeType.STRING,
      },
      tableName: tableNames[`${this.region}/${this.account}`], // Won't work for region-agnostic stacks literally but the point still stands
      replicas: [
        {
          region: 'us-west-2',
        }
      ]
    });
```

But that's not really proper.

So we need a good answer to this. Unresolving this comment.

✅ Retain policies configured
✅ Construct-specific validations passed

Error: Construct upgrade validation failed
Please resolve the drift before proceeding with the upgrade.
For more information about handling drift, visit: 'https://docs.aws.amazon.com/cdk/v2/guide/drift'
```

### Change Set Validation Failure Example

If the change set analysis detects the new table is not being imported, the CLI will show:

```s
Performing construct upgrade validation...

⚠️  Note: This is an experimental feature

Validations:
✅ No CloudFormation drift detected
❌ Change set analysis failed
  │ Resources: MyGlobalTable794EDED1 (AWS::DynamoDB::GlobalTable)
  │ Change Set:
  └  • Action: 'Add' (expected: 'Import')
✅ Retain policies configured
✅ Construct-specific validations passed

Error: Construct upgrade validation failed
```

### Retain Policies Failure Example

Here's the error output example for when retain policies validation fails:

```s
Performing construct upgrade validation...

⚠️  Note: This is an experimental feature

Validations:
✅ No CloudFormation drift detected
✅ Change set analysis complete
❌ Retain policies validation failed
  │ Resource: MyTable794EDED1 (AWS::DynamoDB::Table)

>   └   • RemovalPolicy: 'Delete' (expected: 'Retain')

Reviewer 2: If the removal policy is indeed `Delete`, won't the changeset already tell us that by marking it as a removal? or will the changeset always mark it as removal, not taking into account the policy? 

I'm just wondering if we really need this explicit validation, or can we extract the same information from the changeset.

✅ Construct-specific validations passed

Error: Construct upgrade validation failed
```

#### Construct-specific Validation Failure Example

If construct-specific validation fails, the CLI will show:

```s
Performing construct upgrade validation...

⚠️  Note: This is an experimental feature

Validations:
✅ No CloudFormation drift detected
✅ Change set analysis complete
✅ Retain policies configured
❌ Construct-specific validations failed
  │ Resources: MyGlobalTable794EDED1 (AWS::DynamoDB::GlobalTable)  │ 

>   │ Property 'TableName' is required for import but was not specified

Reviewer 2: If `TableName` is not configured, then change set analysis will fail - do we really need this validation then? will we even get here? It seems more like a cause rather than its own validation. 

Also please add the `SkipReplicationDeletion` validation.

- **Original Author(s):**: @GavinZZ
- **Tracking Issue**: #715
- **API Bar Raiser**: @iliapolo

A construct upgrade framework in the CDK CLI that helps developers upgrade from deployed and
older construct libraries to modern, fully supported alternatives that incorporate the latest
features and best practices.

## Motivation

As the AWS Cloud Development Kit (CDK) evolves, we continuously deliver new features,
performance optimizations, and reliability improvements. To maintain backward compatibility
while enabling forward progress, the CDK team uses structured versioning strategies for critical updates.

A common pattern involves releasing versioned constructs or modules (e.g., V2 variants)
that coexist with their predecessors. For example:

* The `@aws-cdk/aws-ec2-alpha` alpha module introduces the modernized `VpcV2`, `SubnetV2`, and other
new construct, designed to succeed the original Vpc related constructs in `aws-cdk-lib/aws-ec2`.
* The `TableV2` construct in `aws-cdk-lib/aws-dynamodb` replaces legacy Table implementations
while retaining API compatibility.

These updates unlock enhanced capabilities. Construct upgrade between variants, however, often requires

> code changes that could trigger CloudFormation resource replacements (e.g., NAT gateways being

Author: The full workthrough example below should show this, but I can add a small section here.

migration guide format.

- A new CDK CLI command `cdk construct-upgrade --unstable=construct-upgrade` to validate code changes and
execute deployments safely. For more details on the CLI usage, please visit [CLI Settings](#settings)

```sh
cdk construct-upgrade --unstable=construct-upgrade [stacks] \
  --target [id] \
  [OPTIONS]
```

With the initial release of the CDK construct upgrade tool, developers would use the migration guide to
update their CDK code, replacing legacy constructs (e.g., Table) with modern equivalents (e.g., TableV2).

After updating the code to use the new construct, users would use the CDK CLI command `cdk construct-upgrade --unstable=construct-upgrade`
to validates the existing CDK stack and environment, ensuring they are ready for migration. The tool flags any
blockers, such as CloudFormation drift. Additionally, it analyzes the CloudFormation change set and validates
any library-specific configurations. See [Construct Upgrade Validations](#construct-upgrade-validations) for a detailed breakdown of the validations.

In the initial release of the CDK construct upgrade tool, VPC-related constructs upgrade from `aws-cdk-lib/aws-ec2`

> to `@aws-cdk/aws-ec2-alpha` and `Table` to `TableV2` in `aws-cdk-lib/aws-dynamodb` come with enhanced

Reviewer 3: Is moving from a stable construct to an alpha construct really an "upgrade"?

Author: We already have a plan to stablize this before end of Q2, so hopefully it's a stablized module when the framework actually launches.
But IMO, I think this is a good point being raised that and I think people in general won't upgrade to an experimental module given that it may not be backward compatible.

Reviewer 4: Counter opinion: Early adopters will still migrate from current to alpha, if the alpha provides important new features.

⚠️  Note: This is an experimental feature

Validations:
✅ No CloudFormation drift detected
❌ Change set analysis failed
  │ Resources: MyGlobalTable794EDED1 (AWS::DynamoDB::GlobalTable)
  │ Change Set:
  └  • Action: 'Add' (expected: 'Import')

Error: Construct upgrade validation failed
```

**Example 2:** If the change set analysis detects the existing table does not have policy action `Retain` (by setting
the `RemovalPolicy` on existing table construct), the CLI will show:

```s
Performing construct upgrade validation...

⚠️  Note: This is an experimental feature

Validations:
✅ No CloudFormation drift detected
❌ Change set analysis failed
  │ Resource: MyTable794EDED1 (AWS::DynamoDB::Table)
  └   • RemovalPolicy: 'Delete' (expected: 'Retain')

Error: Construct upgrade validation failed
```

**Example 3:** If the change set analysis detects the Replica custom resource does not have policy action `Retain`
(by setting the `ReplicaRemovalPolicy` on existing table construct), the CLI will show:

```s
Performing construct upgrade validation...

⚠️  Note: This is an experimental feature

Validations:
✅ No CloudFormation drift detected
❌ Change set analysis failed

>   │ Resources: MyTableReplicauswest285A33668 (Custom::DynamoDBReplica)

Reviewer 2: Did you validate that retaining the custom resource while removing the backing lambda function is possible? CloudFormation doesn't reject such a change?

  └   • RemovalPolicy: 'Delete' (expected: 'Retain')

Error: Construct upgrade validation failed
```

> ### Custom Validations (Optional Future Enhancement)

Reviewer 5: Isn't this already covered by the support to policy-as-code in the CLI?

Reviewer 4: I like this idea. Would it even be more feasible to implement the construct validations using policy-as-code?

Author: Can you help me understand this a bit more on this policy-as-code feature?

While the currently supported constructs don't require custom validations, future upgrades
(e.g., for AWS::EKS resources) may benefit from this feature. Additionally, this functionality would allow
customers to incorporate enterprise-specific requirements into the validation process.

For instance, users upgrading to TableV2 might want to ensure their existing table construct has
`DeletionProtection` enabled. To accommodate such scenarios, the `cdk construct-upgrade` CLI command would
support custom analysis files as input.  For more details about the setup and file format for custom
validations, refer to the [Custom Validations](#custom-validations-optional-future-enhancement) section.

Example usage:

```sh
cdk construct-upgrade --unstable=construct-upgrade --target aws-cdk-lib.aws-dynamodb.TableV2 \ 

>   --custom-analysis table-custom-analysis.ts

Reviewer 3: Woah! People are going to submit source code?

```

If the custom validation detects that `DeletionProtection` is not enabled on existing table construct,
the CLI would display:

```s
Performing construct upgrade validation...

⚠️  Note: This is an experimental feature

Validations:
✅ No CloudFormation drift detected
✅ Change set analysis complete
❌ Custom validation failed
  │ Resource: MyTable794EDED1 (AWS::DynamoDB::Table)
  └   • DeletionProtection: 'false' (expected: 'true')

Error: Construct upgrade validation failed
```

This feature would enhance the flexibility of the construct upgrade process, allowing teams to enforce
specific policies or best practices during migrations.

> If you're interested in seeing a full example of VPC V2 example, please checkout
> [Appendix B: VPC V2 Upgrade Example](#appendix-b-vpc-related-migration-example).

> ### User Experience: Visual Workflow

Reviewer 6: If you are going to show a visual with named blocks, I think you need to explain what each block is before or provide a key with links to where the explanation is. Ensure explanation covers both why it is needed and how it works.

Author: Thanks Simon, I'll move it to the later section after each individual validations are explained.

| │   Human refactor    │       │Other refactor method│     |
| │                     │       │      like LLM       │     |
| └─────────────────────┘       └─────────────────────┘     |
|            │                             │                |
|            └──────────────┬──────────────┘                |
|                           |                               |
└───────────────────────────|───────────────────────────────┘
                            |
┌───────────────────────────|───────────────────────────────┐
|                Automated Workflow by CDK CLI              | 
|                           |                               |
|                           ▼                               |
|                ┌─────────────────────┐                    |
|                │ Validate CFN drift  │                    |
|                └─────────────────────┘                    |
|                           |                               |
|                           ▼                               |
|                ┌─────────────────────┐                    |
|                │   CFN Change set    |                    |

> |                |       analysis      │                    |

Reviewer 3: Not explicitly set, but we create the change set here I suppose?

A change set that would just try to deploy the new template? And in the case of the DynamoDB table migration, we would see:

* **Delete** `AWS::DynamoDB::Table` with dynamic name
* **Add** `AWS::DynamoDB::GlobalTable` with name `DemoStack-MyTable794EDED1-11W4MR8VZ0UPE`

And our validation would say "yes that is what we want to see" ?

Author: A change set does not deploy the new template until we execute it.

In the case of the DynamoDB table migration, we would see:

- Remove AWS::DynamoDB::Table (with policyAction saying 'Retain')
- Add AWS::DynamoDB::GlobalTable with name DemoStack-MyTable794EDED1-11W4MR8VZ0UPE and the corresponding action is 'Import'

|| |    refactoring    | |         || Retain-Remove-Import||| 
|| └───────────────────┘ |         |└─────────────────────┘|| 
|└───────────────────────┘         └───────────────────────┘|
|           │                               │               | 
|           └───────────────────────────────┘               |
|                          |                                |
|                          ▼                                |
|               ┌─────────────────────┐                     |
|               │        Done         │                     |
|               └─────────────────────┘                     |
└───────────────────────────────────────────────────────────┘
```

## Settings

You have a few settings available to control the behavior of the new CDK CLI command
for construct upgrade feature.

```sh
cdk construct-upgrade --unstable=construct-upgrade [stacks] \
  --target [id] \
  [OPTIONS]
```

Arguments:

- CDK stack ID (Optional)
    The construct ID of the CDK stack from your app to synthesize.

Options:

- --target (Required):
  The target construct and module you want to upgrade to. For example, `aws-cdk-lib.aws-dynamodb.TableV2` or `@aws-cdk/aws-ec2.VpcV2`

>   For fully supported construct upgrades, we can short hand it to `TableV2` or `VpcV2`.

Reviewer 3: Shouldn't this identify an instance (construct)? Why does it identify a class?

Author: This way we allow simultaneous multiple instances of the construct to upgrade at the same time. In general, it's up to users' code change, but if users update multiple Vpcs to VpcV2s, I think it makes sense for the construct upgrade framework to upgrade it all.

- --dry-run (Optional)
  This flag is default to false. When the flag is set to true, the CLI acts as dry-run and will only trigger validation.

> - --require-approval (Optional)

Reviewer 3: Nit, but this flag is poorly named. It should have been called `--require-confirmation`. We should deprecate the old flag, add the new one, and don't add any more copies of the silly name.

  Specify what security-sensitive changes require manual approval.
  - any-change – Manual approval required for any change to the stack.

>   - broadening – Manual approval required if changes involve a broadening of permissions or security group rules.

Reviewer 5: Do we expect this case to happen?

  - never – Approval is not required.
  Valid values: [any-change, broadening, never]
  Default value: broadening
- --skip-validation (Optional)
  Default to false. Use as users' own risk to by pass the validations step.
- --custom-analysis (Optional)
  When user provides a custom JavaScript file for validation, more will be explained in [Custom Validations](#custom-validations-optional-future-enhancement)
- Other global options like --quiet, --help, etc

## Construct Upgrade Strategies

The CDK construct upgrade framework employs two different strategies for migrating resources, automatically
selecting the appropriate approach based on the `--target` construct. CDK CLI stores and uses a context variable
to determine the mapping between upgrade targets and their corresponding strategies.

### In-place Migration

This strategy is suitable for constructs where changes do not involve custom resource
removal or introduction of new underlying resource types. It leverages the CloudFormation stack refactoring

> feature to maintain resource continuity.

Reviewer 6: This could be clearer on the problem it solves, "maintain resource continuity" is a bit vague. Today moving constructs around and similar changes in a CDK application can cause the logical id of the resulting cloud-formation resource to change, resulting in a deletion and recreation of the resources. Stack refactoring enables a logical name change without any change to the deployed resource.

Currently, users must specify a `refactor.json` file that maps logical ID changes from the old construct to the new one.
If construct upgrade validations are successful, the command continues to execute and deploy the CDK stack using CloudFormation
stack refactor feature, which helps retain existing resources while applying the necessary infrastructure updates.

> > Future Enhancement: CDK team is developing a [built-in refactor feature](https://github.com/aws/aws-cdk-rfcs/pull/705)

Reviewer 5: The RFC has already been merged:

```suggestion
> Future Enhancement: CDK team is developing a [built-in refactor feature](https://github.com/aws/aws-cdk-rfcs/blob/main/text/0162-refactoring-support.md)
```

> that will eliminate the need for manual refactor.json files. The mapping will be automatically inferred
> from the construct tree and paths. Once implemented, this process will be fully automated.

### Retain-Remove-Import Migration

This strategy is employed when significant changes occur in the underlying resource, such as transitioning from
a custom resource to a native L1 resource or resource property (e.g., Table to TableV2 migration). The CDK construct upgrade
framework handles this process through several automated steps:

- Validation of `RemovalPolicy.RETAIN` on the original construct to prevent accidental resource deletion.
- Execution of the `cdk construct-upgrade` command, which performs necessary validations.

> - If validations pass, automatic triggering of deployment using CDK's import feature to link existing resources

Reviewer 3: Missing:

- A deployment that removes the resources from the template.

> provide information about changes and is designed to be status-agnostic.

Sample output of `cdk diff --detect-drift` command is as follows:

```sh
$ cdk diff --detect-drift

Drifts
[~] AWS::Lambda::Function MyFunction drifted
[~] AWS::DynamoDB::Table MyTable drifted

Resources
[-] AWS::DynamoDB::Table MyTable orphan
[+] AWS::DynamoDB::GlobalTable MyTable import
```

The new `cdk construct-upgrade` CLI command will implicitly call the `cdk diff --detect-drift` command as part of
its validation process. To avoid being overly restrictive, the validation will primarily focus on the main resources
change. The main resources can be determined based on the `--target` option, more can be found in
[Appendix C: Main Resources](#appendix-c-main-resources-computation).

Drift detection is important to ensure construct upgrade safety by preventing accidental deployment issues when
the CDK construct state does not match the actual deployed resource state. Without detecting drift, upgrades may
unintentionally remove manual configurations made through the AWS Console, potentially triggering resource

> replacements, data loss, or availability downtime.

Reviewer 3: Note that the only way to "resolve" drift manually is to manually undo those manual configurations via the AWS Console.

Changing your template to match reality may or may not do the right thing (it may even fail to deploy), and so it's unreliable and we cannot recommend it.

```yaml
- cloudformation:DetectStackDrift
- cloudformation:DetectStackResourceDrift
- cloudformation:DescribeStackDriftDetectionStatus
```

To add these required permissions, run the following command target environment:

```sh
cdk bootstrap
```

This step is crucial as it will configure the environment with the correct permissions,
enabling the validation and migration process to proceed smoothly.

### CloudFormation Change Set Analysis

Another validation is to validate CloudFormation change set diff. The current `cdk diff` command implicitly calls
CloudFormation change set creation, providing high-level details such as "add", "delete", "modify", "import",
and etc. like the following:

```s
[-] AWS::DynamoDB::Table MyTable orphan
[+] AWS::DynamoDB::GlobalTable MyTable add
```

For `In-Place Migration` strategy, change set analysis is not required as CloudFormation stack refactoring enforces that
the refactor operations don't allow new resource creations, resource deletions, or changes to resource configurations,
see [Stack refactoring limitations](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/stack-refactoring.html#stack-refactoring-limits).

> For the `Retain-Remove-Import Migration` strategy, we will introduce an additional option `--import-existing-resources`

Reviewer 5: I find all this part until the end of the section hard to follow. There are a lot of implementation details, but I don't understand what is important from the user's point of view.

Here is some future enhancements that we would make to the custom validations, see
[Appendix E: Custom Validation Future Enhancement](#appendix-e-custom-validation-future-enhancements).

## CI/CD Integration

Enterprise developers often lack production AWS credentials, making pre-deployment
validations (e.g., drift detection, change set analysis) impossible in production environments.

At development time, here is how it works for common enterprise scenarios. Assuming developers have upgrade
their CDK stack code to use new construct. They can use `dry-run` feature which will allow users to
run the validations on the test/dev account without deploying.

```sh
$ cdk construct-upgrade --unstable=construct-upgrade --target aws-cdk-lib.aws-dynamodb.TableV2 --dry-run
Performing construct upgrade validation...

⚠️  Note: This is an experimental feature

Validations:
✅ No CloudFormation drift detected
✅ Change set analysis complete
```

Depending on the deployment workflow and environment constraints, there are two common enterprise
patterns:

1. CI/CD-driven Deployment:
For teams with access to CI/CD and production environments, developers commit the upgraded CDK stack

> code to version control. The CI/CD pipeline runs `cdk construct-upgrade` as part of the deployment process.

Reviewer 3: Note that this explicitly won't work for deployment pipelines as used by Amazon, since they deploy CloudFormation stacks only and don't run arbitrary CLI commands.

Alternatively, the development team can upgrade the source code, commit and push the changes with CI/CD pipelines
deployment temporarily disabled. The operations team runs the migration in each production region and validates
deployment success. The development team re-enables CI/CD to roll out the changes through the normal pipeline.

**Known Limitation**: CI/CD-driven deployment has limitations when using pipeline systems that don't easily allow
command configurations (e.g., CDK L3 Pipeline construct where deployments rely on CloudFormation actions rather
than direct CDK CLI commands).

In these systems, implementing the CI/CD-driven deployment would require:

- Blocking the CI system while adding the `cdk construct-upgrade` command
- Committing the change to trigger the upgrade validation
- Unblocking the CI system and preventing new commits during the process
- Removing the upgrade command from CI pipeline configuration after completion

This process is cumbersome and typically requires administrative permissions to modify the CI system. Given
these challenges, we recommend using Ops-led deployment for teams with complex pipeline setups until a more

> elegant solution is developed in the future updates.

Reviewer 7: I think it's really important to define _complex_ here. It would also be good to understand how many customers have complex setups i.e. does the proposed approach actually work for a majority of customers?

Author: We don't currently have a way to track the analytics usage of users' CI/CD pipeline setup unfortunately. That's why we provided a workaround approach for a less-ideal solution.

> ## Rollback

Reviewer 2: What about if the user wants to go back to the previous variant? not because it failed - just because they changed their minds. 

Will running `cdk construct-upgrade aws-cdk-lib.aws_dynamodb.Table` work for the reverse migration?

Author: Yes, it will work smoothly. I can add some more content to Rollback section to explain this. The same can also be applied to the refactor approach.

to detect that users are attempting a construct upgrade and prevent potentially destructive changes.

A future enhancement would add detection capabilities to the `cdk deploy` command. The CLI would analyze the change
set prior to deployment and compare it against the context file mappings (See [Appendix C: Main Resources](#appendix-c-main-resources-computation)):

```json
{
  "aws-cdk-lib.aws_dynamodb.TableV2": {
    "strategy": "Import",
    "source": ["AWS::DynamoDB", "Custom::ReplicationDeletion"],
    "target": ["AWS::DynamoDB::GlobalTable"]
  }
}
```

If the CLI detects resource removals matching the source types and resource additions matching the target types, it would either:

- Display a warning message about the potential construct upgrade scenario and prompt users to use `cdk construct-upgrade`
  and the corresponding migration guide, or
- Present a confirmation prompt requiring explicit user approval to proceed (unless `--require-approval never` is specified).

This enhancement would help prevent accidental resource replacements by guiding users toward the proper construct upgrade process.

### Limitation: In-Place Migration Strategy

> This strategy utilizes the CDK/CloudFormation refactoring feature and inherits its limitations. It experiences

Author: Not sure what you meant here, the alternative is to use CDK import feature in the other strategy.

the [same limitations](https://github.com/aws/aws-cdk-rfcs/blob/otaviom/refactoring-support/text/0162-refactoring-support.md#pipelines-with-version-superseding)
as described in the CDK Refactoring RFC. Primarily, these limitations affect pipelines with version superseding,

> potentially causing issues in complex deployment scenarios.

Reviewer 7: Would be great to understand the complex deployment scenarios and how likely they are to cause issues.

Author: The phrasing is inaccurate, it's actually talking about the issue right above where the pipeline can't customize the CLI command to run.

### Limitation: Retain-Remove-Import Migration Strategy

This strategy relies on the CDK import feature, which requires importing resources through their physical names.
For DynamoDB tables, users must define the `TableName` property in the construct. In CI/CD pipelines, where physical
names vary across accounts and regions, users need to maintain a JSON file mapping accounts to table physical names.

1. Create a JSON file (e.g., table-names.json) with the physical names:

```json
{
  "dev": "DevStack-MyTable794EDED1-11W4MR8VZ0UPE",
  "prod": "ProdStack-MyTable794EDED1-22X5NS9WA1VPF"
}
```

2. Read from this file during stack synthesis:

```ts
const tableNames = JSON.parse(fs.readFileSync('./table-names.json', 'utf8'));

const table = new dynamodb.TableV2(this, 'MyTable', {
  // ... other properties ...
  tableName: tableName['prod'],
});
```

> > Important: The tableName property is required for the upgrade process and cannot be removed afterward, as CloudFormation will

Reviewer 2: We spoke about possibly enhancing the `deploy` command to support a mapping file, just like `import` does:

```console
cdk deploy --resource-mapping /path/to/mapping.json
``` 

Doing so will allow omitting the `tableName` property from the CDK code itself, and thus not introduce this limitation. Does this paragraph mean to say that `cdk deploy` cannot be enhanced like that? If so - i'm not sure I see why. Can you elaborate?

Author: It's possible but this will cause the whole workflow to be very complex.

The resourcesToImport features does not work unfortunately in this case because what it does requires the stack to be identical except the newly created resource that will be imported. What this means is that the user workflow will be torn into multiple step:

    Users add a new TableV2 construct
    Users run cdk import/cdk deploy (with migrate.json to use resourcesToImport feature).
    Once import is successful, users remove the TableV1 construct
    Deploy the change

This now becomes a multi-step and non-automated step with user intevention in between. I do not like this approach and I don't think it's worth the pain. I outlined a few possible workarounds instead to not hardcode the value in the stack.

    1. Verify if custom resource deletion does not delete the replica tables (through a L2 prop)
    2. Verify if properties like `TableName` for DynamoDB table construct is specified and Replica property exists
    3. Verify if properties like `DeletionProtection` for DynamoDB table is configured for safety

##### Phase 2 (execution)

Deploy the migrated stack using the migration tool if validations are completed successfully.
This phase differs depends on the module being migrated due to different migration approaches for Table and VPC.

High-level tasks:

1. Support In-Place migration execution
2. Support Retain-Remove-Import migratione execution

#### Milestone 2

Milestone 2 will be a future release to help automate the code migration process as well.
This is still under investigation and may or may not be included in the CDK construct upgrade support.

### Will this tool support every future V1 to V2 construct?

The CDK construct upgrade tool is designed to support common and high-impact migration scenarios for widely adopted
constructs and modules. While we aim to expand coverage over time, we cannot guarantee support for every

> future V1-to-V2 migration path due to variations in technical complexity and resource constraints.

Reviewer 8: How will users know which constructs are supported? Will there be clear documentation?

Author: We'll publish documentation around the CLI usage and the `--target` value should enumerate the valid values.

    const publicSN = new ec2alpha.SubnetV2(this, PUBLIC_SUBNET_NAME, {
      vpc: this.vpc,
      availabilityZone: AZ,
      ipv4CidrBlock: new ec2alpha.IpCidr(`10.0.0.0/24`),
      subnetType: ec2.SubnetType.PUBLIC,
      mapPublicIpOnLaunch: true,
      subnetName: PUBLIC_SUBNET_NAME,
    });
    cdk.Tags.of(publicSN).add('Name', `${this.stackName}/vpc/public-subnet-test-cyberark1Subnet1`);

    new ec2alpha.Route(this, 'private-route', {
      routeTable: privateSN.routeTable,
      destination: "0.0.0.0/0",
      target: new ec2alpha.RouteTargetType({ gateway: ntgwy })
    });
    cdk.Tags.of(privateSN.routeTable).add('Name', `${this.stackName}/vpc/private-subnet-test-cyberark1Subnet1`);

    new ec2alpha.Route(this, 'public-route', {
      routeTable: publicSN.routeTable,
      destination: "0.0.0.0/0",
      target: new ec2alpha.RouteTargetType({ gateway: internetgw })
    });
    cdk.Tags.of(publicSN.routeTable).add('Name', `${this.stackName}/vpc/public-subnet-test-cyberark1Subnet1`);
  }
}
```

Now, validate the construct upgrade using the CDK CLI:

> Note that the flag `--unstable=construct-upgrade` is necessary to indicate that

> > it's an experiemntal feature that users are using until feature stablized.

Reviewer 7: [Possibly out of scope for the RFC] What does it take for the feature to be stabilised?

Reviewer 4: We don't have a strict criteria for that, especially not for CLI features. Generally the idea is to give us and users some bake time (at least 3 months, better 6 months) to figure out any issues and papercuts that might require breaking changes to resolve. E.g. renaming or removing options.

Author: Good question, I think this is more of a question to be answered by the core team. Can someone from core team chime in and help explain the general process to stablize a CLI command?

```sh

> cdk construct-upgrade --unstable=construct-upgrade --target @aws-cdk.aws-ec2-alpha.VpcV2

Reviewer 2: Its somewhat confusing to use `@aws-cdk.aws-ec2-alpha.VpcV2` as the target here because the code change actually introduces many more new concstructs:

- ec2alpha.Route
- ec2alpha.SubnetV2
- ...

Can we think of a different value here that more accurately describes the migration? Maybe:

```suggestion
cdk construct-upgrade --unstable=construct-upgrade --target @aws-cdk.aws-ec2-alpha
```

?

  │ Resource: test-public-subnet794EDED1 -> vpc-test-public-subnet1234FBDA
  └ Error: Resource configuration mismatch

Error: Construct upgrade validation failed
Please resolve the drift before proceeding with the upgrade.
For more information about handling drift, visit: 'https://docs.aws.amazon.com/cdk/v2/guide/drift'
```

### Appendix C: Main Resources Computation

As described in drift detection validation and change set validation, to avoid being overly restrictive, the validation
will primarily focus on the main resources associated with the construct upgrades.

The main resources are determined using two inputs:

- The required `--target` option specified in the construct upgrade CLI command (e.g., `cdk construct-upgrade --target aws-cdk-lib.aws_dynamodb.TableV2`)
- A pre-defined context file stored in the CDK codebase

CDK stores a JSON formatted context file in the CDK CLI codebase that maps targets to their upgrade configurations:

```json
{
  "aws-cdk-lib.aws_dynamodb.TableV2": {
    // Valid values are 'Import' and 'Refactor'
    "strategy": "Import",
    
    // Resource type prefixes to check in drift detection and retain policy validation

>     "source": ["AWS::DynamoDB", "Custom::ReplicationDeletion"],

Reviewer 2: Can we be more accurate here?

```suggestion
    "retain": ["AWS::DynamoDB::Table", "Custom::ReplicationDeletion"],
```

Also, this list is relevant only for the `aws-cdk-lib.aws_dynamodb.Table` construct. It means we are making an assumption that the upgrade is always from `Table` to `TableV2`. 

This holds true because those are the only two we have **right now**, but once a future `TableV3` is introduced, this context becomes ambiguous:

```json
{
  "aws-cdk-lib.aws_dynamodb.TableV3": {...}
}
```

What does the above context refer to? is it `Table -> TableV3` OR `Tablev2 -> TableV3`. Nothing guarantees that the same information is sufficient for both. Since this is an implementation detail, we can evolve it however we want to fit that use case, so it doesn't need to be addressed now; but it would be good to see how such context will support upgrade when more than 2 variants are available.

#### Thought for the future

This list is essentially **the list of stateful resources** this construct encompasses. It can and should be at some point available as static data that the CLI has access to. This will allow `cdk deploy` to optionally block removal of stateful resources regardless of whether an upgrade or a regular deploy is performed.

    // Resource types to verify import action in change set analysis

>     "target": ["AWS::DynamoDB::GlobalTable"]

Reviewer 2: ```suggestion
    "import": ["AWS::DynamoDB::GlobalTable"]
```

  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const table = new dynamodb.Table(this, 'MyTable', {
      partitionKey: {
        name: 'PK',
        type: dynamodb.AttributeType.STRING,
      },
      replicationRegions: ['us-west-2'],
    });
  }
}
```

**Step 1:**: CDK users follow the construct upgrade guide provided by CDK team and upgrade to TableV2 construct:

```ts
import * as dynamodb from 'aws-cdk-lib/aws-dynamodb';

export class DemoStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const table = new dynamodb.TableV2(this, 'MyTable', {
      partitionKey: {
        name: 'PK',
        type: dynamodb.AttributeType.STRING,
      },
      tableName: 'DemoStack-MyTable794EDED1-11W4MR8VZ0UPE',
      replicas: [
        {
          region: 'us-west-2',
        }
      ]
    });
  }
}
```

**Step 2:**: CDK users follow the construct upgrade guide to validate code changes by running the
CDK CLI commands below:

```s
# exit with 1 if CloudFormation drift is detected for resource type AWS::DynamoDB::Table
# exit with 0 otherwise

> $ cdk drift --rejectAWS::DynamoDB::Table

Reviewer 3: Why do we need to reject specifically this? Would we not reject on any drift at all?

(I mean, it's fine to have options on the command if a user knows they want to ignore drift... but it shouldn't [need to be] be part of the **expected** flow)

Stack DemoStack
✅ No CloudFormation drift detected

# exit with 1 if resources types AWS::DynamoDB::Table and Custom::DynamoDBReplica are deleted, OR
# exit with 1 if resources type AWS::DynamoDB::GlobalTable are not imported
# exit with 0 otherwise

> $ cdk diff --retain AWS::DynamoDB::Table,Custom::DynamoDBReplica --import AWS::DynamoDB::GlobalTable

Reviewer 3: This is an annoying command line to type.

I will once again float my idea of doing `cdk diff --validate-migration=<migration plan name>`.

I think the migration plan should also **print to the user** the command they're expected to run next.

Stack DemoStack
✅ There were no differences
```

**Step 3:**: CDK users follow the construct upgrade guide to deploy the changes by running the
CDK CLI command below:

```s

> $ cdk deploy --import-existing-resources

Reviewer 3: How about `cdk refactor --retain-delete-import`, which it definitely is going to have to be for this specific migration.

The workflow is then always 

1. Update code
2. `cdk diff --validate-whatever`
3. `cdk refactor`

At development time, here is how it works for common enterprise scenarios. Assuming developers have upgrade
their CDK stack code to use new construct. They can run the following commands to validate changes.

```sh
$ cdk drift --rejectAWS::DynamoDB::Table
$ cdk diff --retain AWS::DynamoDB::Table,Custom::DynamoDBReplica --import AWS::DynamoDB::GlobalTable
```

Depending on the deployment workflow and environment constraints, there are two common enterprise
patterns:

1. **CI/CD-driven Deployment**:
For teams with access to CI/CD and production environments, developers commit the upgraded CDK stack
code to version control. The CI/CD pipeline runs the following as part of the deployment process:

```sh
# Run validations and store exit code
cdk drift --rejectAWS::DynamoDB::Table && cdk diff --retain AWS::DynamoDB::Table,Custom::DynamoDBReplica --import AWS::DynamoDB::GlobalTable

if [ $? -eq 0 ]; then
    echo "Validation passed. Proceeding with deployment..."
    cdk deploy --import-existing-resources # `cdk refactor --unstable=refactor` for VPC constructs upgrade
else
    echo "Validation failed. Deployment blocked."
    exit 1
fi

> ```

Reviewer 3: Can you be specific about the entire workflow end to end? I think it looks like this:

1. Clear their deployment pipelines, no change may be flowing through it
2. Update their pipeline definition to put the above command in it
3. Deploy a single change, the migration change
4. Wait for the change to clear the pipeline (no other changes may go through it)
5. Update their pipeline definition to remove the above command

Is that right?

----

Now think of a typical AWS pipeline that has ~3 pre-prod stages and >20 prod stages, and various approval workflows. Does the above still sound acceptable?

Reviewer 3: I think we should say "CI/CD" is too hard to design right now, we will do it later.

Author: I think CI/CD is one of the core feature for construct upgrade. We don't have to fully support complex setup but we should definitely explain the workflow for simpler setup.

I agree the workflow you mentioned doesn't seem ideal especially when users want to intermix construct upgrade with normal feature release changes. The issue with AWS pipeline is that there's usually a 12 hours bake time in between each stage which makes things worse. In either these two case,s I think we can add a note to the RFC to explain the difficulty and suggest users to use the Ops-Lead workflow instead.
